File: Chapter12EnvironmentSensorSound/LoadSensorExample.java
package X3dForWebAuthors.Chapter12EnvironmentSensorSound;

/*
Copyright (c) 1995-2022 held by the author(s).  All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer
      in the documentation and/or other materials provided with the
      distribution.
    * Neither the name of the Web3D Consortium (https://www.web3d.org)
      nor the names of its contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
*/

import org.web3d.x3d.jsail.Core.*;
import org.web3d.x3d.jsail.fields.*;
import org.web3d.x3d.jsail.Grouping.*;
import org.web3d.x3d.jsail.Interpolation.*;
import org.web3d.x3d.jsail.Navigation.*;
import org.web3d.x3d.jsail.Networking.*;
import org.web3d.x3d.jsail.Scripting.*;
import org.web3d.x3d.jsail.Shape.*;
import org.web3d.x3d.jsail.Text.*;
import org.web3d.x3d.jsail.Time.*;

// Javadoc annotations follow, see below for source.
/**
 * <p> Simple test of LoadSensor node. </p>
 <p> Related links: <a href="../../../Chapter12EnvironmentSensorSound/LoadSensorExample.java">LoadSensorExample.java</a> source, <a href="../../../Chapter12EnvironmentSensorSound/LoadSensorExampleIndex.html" target="_top">LoadSensorExample catalog page</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dResources.html" target="_blank">X3D Resources</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dSceneAuthoringHints.html" target="_blank">X3D Scene Authoring Hints</a>, and <a href="https://www.web3d.org/x3d/content/X3dTooltips.html" target="_blank">X3D Tooltips</a>. </p>
	<table style="color:black; border:0px solid; border-spacing:10px 0px;">
        <caption>Scene Meta Information</caption>
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center; padding:10px 0px;"><i>meta tags</i></td>
			<td style="text-align:left;   padding:10px 0px;">&nbsp; Document Metadata </td>
		</tr>

		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> title </i> </td>
			<td> <a href="../../../Chapter12EnvironmentSensorSound/LoadSensorExample.x3d">LoadSensorExample.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> description </i> </td>
			<td> Simple test of LoadSensor node </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> creator </i> </td>
			<td> Don Brutzman </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> created </i> </td>
			<td> 17 July 2008 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> modified </i> </td>
			<td> 14 June 2020 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> rights </i> </td>
			<td> Copyright Don Brutzman and Leonard Daly 2008 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> subject </i> </td>
			<td> X3D LoadSensor example </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> identifier </i> </td>
			<td> <a href="https://X3dGraphics.com/examples/X3dForWebAuthors/Chapter12EnvironmentSensorSound/LoadSensorExample.x3d" target="_blank">https://X3dGraphics.com/examples/X3dForWebAuthors/Chapter12EnvironmentSensorSound/LoadSensorExample.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> generator </i> </td>
			<td> X3D-Edit 3.3, <a href="https://savage.nps.edu/X3D-Edit" target="_blank">https://savage.nps.edu/X3D-Edit</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> license </i> </td>
			<td> <a href="../../../Chapter12EnvironmentSensorSound/../license.html">../license.html</a> </td>
		</tr>
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center;" colspan="2">  &nbsp; </td>
		</tr>
	</table>

	<p>
		This program uses the
		<a href="https://www.web3d.org/specifications/java/X3DJSAIL.html" target="_blank">X3D Java Scene Access Interface Library (X3DJSAIL)</a>.
		It has been produced using the 
		<a href="https://www.web3d.org/x3d/stylesheets/X3dToJava.xslt" target="_blank">X3dToJava.xslt</a>
		stylesheet to create Java source code from an <code>.x3d</code> model.
	</p>

	* @author Don Brutzman
 */

public class LoadSensorExample
{
	/** Default constructor to create this object. */
	public LoadSensorExample ()
	{
	  initialize();
	}

	/** Create and initialize the X3D model for this object. */
	public final void initialize()
	{
  x3dModel = new X3D().setProfile(X3D.PROFILE_IMMERSIVE).setVersion(X3D.VERSION_3_3)
  .setHead(new head()
    .addMeta(new meta().setName(meta.NAME_TITLE      ).setContent("LoadSensorExample.x3d"))
    .addMeta(new meta().setName(meta.NAME_DESCRIPTION).setContent("Simple test of LoadSensor node"))
    .addMeta(new meta().setName(meta.NAME_CREATOR    ).setContent("Don Brutzman"))
    .addMeta(new meta().setName(meta.NAME_CREATED    ).setContent("17 July 2008"))
    .addMeta(new meta().setName(meta.NAME_MODIFIED   ).setContent("14 June 2020"))
    .addMeta(new meta().setName(meta.NAME_RIGHTS     ).setContent("Copyright Don Brutzman and Leonard Daly 2008"))
    .addMeta(new meta().setName(meta.NAME_SUBJECT    ).setContent("X3D LoadSensor example"))
    .addMeta(new meta().setName(meta.NAME_IDENTIFIER ).setContent("https://X3dGraphics.com/examples/X3dForWebAuthors/Chapter12EnvironmentSensorSound/LoadSensorExample.x3d"))
    .addMeta(new meta().setName(meta.NAME_GENERATOR  ).setContent("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit"))
    .addMeta(new meta().setName(meta.NAME_LICENSE    ).setContent("../license.html")))
  .setScene(new Scene()
    .addChild(new WorldInfo().setTitle("LoadSensor.x3d"))
    .addChild(new Viewpoint("ViewAll").setDescription("LoadSensor test").setPosition(0.0,0.0,12.0))
    .addChild(new Inline("HelloWorld").setUrl(new String[] {"../HelloWorld.x3d","https://www.web3d.org/x3d/content/examples/HelloWorld.x3d"}))
    .addChild(new LoadSensor("InlineLoadSensor").setTimeOut(5)
      .addChild(new Inline().setUSE("HelloWorld")))
    .addChild(new Transform("PivotTextSigns").setTranslation(0.0,3.0,0.0)
      .addComments(" First sign indicates waiting... ")
      .addChild(new Shape()
        .setGeometry(new Text().setLength(new double[] {0.0}).setString(new String[] {"LoadSensor waiting for HelloWorld scene..."})
          .setFontStyle(new FontStyle().setJustify(FontStyle.JUSTIFY_MIDDLE_MIDDLE).setSize(0.7).setCssStyle("ITALIC")))
        .setAppearance(new Appearance()
          .setMaterial(new Material().setEmissiveColor(0.878431,0.262745,0.0))))
      .addComments(" Second sign indicates completed, initially rotated 90 degrees out of view ")
      .addChild(new Transform().setRotation(0.0,1.0,0.0,1.57)
        .addChild(new Shape()
          .setGeometry(new Text().setLength(new double[] {0.0}).setString(new String[] {"LoadSensor reports HelloWorld scene loading complete."})
            .setFontStyle(new FontStyle().setJustify(FontStyle.JUSTIFY_MIDDLE_MIDDLE).setSize(0.7).setCssStyle("ITALIC")))
          .setAppearance(new Appearance()
            .setMaterial(new Material().setEmissiveColor(0.0,0.278431,0.027451))))))
    .addChild(new OrientationInterpolator("SignRotationInterpolator").setKey(new double[] {0.0,1.0}).setKeyValue(new MFRotation(new double[] {0.0,1.0,0.0,0.0,0.0,1.0,0.0,-1.57})))
    .addChild(new TimeSensor("SignRotationClock").setCycleInterval(0.8))
    .addChild(new ROUTE().setFromNode("InlineLoadSensor").setFromField("loadTime").setToNode("SignRotationClock").setToField("set_startTime"))
    .addChild(new ROUTE().setFromNode("SignRotationClock").setFromField("fraction_changed").setToNode("SignRotationInterpolator").setToField("set_fraction"))
    .addChild(new ROUTE().setFromNode("SignRotationInterpolator").setFromField("value_changed").setToNode("PivotTextSigns").setToField("set_rotation"))
    .addChild(new Script("TraceScript").setSourceCode("""
ecmascript:
function eventLoaded (value)
{
	Browser.println ('InlineLoadSensor isLoaded=' + value);
}
function loadTimeEvent (value)
{
	Browser.println ('InlineLoadSensor loadTime=' + value);
}
""")
      .addField(new field().setName("eventLoaded").setType(field.TYPE_SFBOOL).setAccessType(field.ACCESSTYPE_INPUTONLY))
      .addField(new field().setName("loadTimeEvent").setType(field.TYPE_SFTIME).setAccessType(field.ACCESSTYPE_INPUTONLY)))
    .addChild(new ROUTE().setFromNode("InlineLoadSensor").setFromField("isLoaded").setToNode("TraceScript").setToField("eventLoaded"))
    .addChild(new ROUTE().setFromNode("InlineLoadSensor").setFromField("loadTime").setToNode("TraceScript").setToField("loadTimeEvent")));
    }
	// end of initialize() method

	/** The initialized model object, created within initialize() method. */
	private X3D x3dModel;

	/** 
	 * Provide a 
	 * <a href="https://dzone.com/articles/java-copy-shallow-vs-deep-in-which-you-will-swim" target="_blank">shallow copy</a>
	 * of the X3D model.
	 * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/Core/X3D.html">X3D</a>
	 * @return LoadSensorExample model
	 */
	public X3D getX3dModel()
	{	  
		return x3dModel;
	}
	   
    /** 
	 * Default main() method provided for test purposes, uses CommandLine to set global ConfigurationProperties for this object.
     * @param args array of input parameters, provided as arguments
	 * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/Core/X3D.html#handleArguments-java.lang.String:A-">X3D.handleArguments(args)</a>
	 * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/Core/X3D.html#validationReport--">X3D.validationReport()</a>
     * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/CommandLine.html">CommandLine</a>
     * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/CommandLine.html#USAGE">CommandLine.USAGE</a>
     * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/ConfigurationProperties.html">ConfigurationProperties</a>
     */
    public static void main(String args[])
    {
        X3D thisExampleX3dModel = new LoadSensorExample().getX3dModel();

		boolean hasArguments = (args != null) && (args.length > 0);
		boolean validate = true; // default
		boolean argumentsLoadNewModel = false;
		String  fileName = new String();

		if (args != null)
		{
			for (String arg : args)
			{
				if (arg.toLowerCase().startsWith("-v") || arg.toLowerCase().contains("validate"))
				{
					validate = true; // making sure
				}
				if (arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_X3D) ||
					arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_CLASSICVRML) ||
					arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_X3DB) ||
					arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_VRML97) ||
					arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_EXI) ||
					arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_GZIP) ||
					arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_ZIP) ||
					arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_HTML) ||
					arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_XHTML))
				{
					argumentsLoadNewModel = true;
					fileName = arg;
				}
			}
		}
		if      (argumentsLoadNewModel)
			System.out.println("WARNING: \"X3dForWebAuthors.Chapter12EnvironmentSensorSound.LoadSensorExample\" model invocation is attempting to load file \"" + fileName + "\" instead of simply validating itself... file loading ignored.");
		else if (hasArguments) // if no arguments provided, this method produces usage warning
			thisExampleX3dModel.handleArguments(args);

		if (validate)
		{
			System.out.print("Java program \"X3dForWebAuthors.Chapter12EnvironmentSensorSound.LoadSensorExample\" self-validation test results: ");
			String validationResults = thisExampleX3dModel.validationReport();
            if (validationResults.startsWith("\n"))
                System.out.println();
			System.out.println(validationResults.trim());
		}
    }
}

File: Chapter12EnvironmentSensorSound/LoadSensorExample.py
####################################################################################################
#
# Now available: developmental python x3d.py package on PyPi for import.
#   This approach simplifies Python X3D deployment and use.
#   https://pypi.org/project/x3d
#
# Installation:
#       pip install x3d
# or
#       python -m pip install x3d
#
# Developer options for loading x3d package:
#
#    from x3d import *  # preferred approach, terser source that avoids x3d.* class prefixes
#
# or
#    import x3d         # traditional way to subclass x3d package, all classes require x3d.* prefix,
#                       # but python source is very verbose, for example x3d.Material x3d.Shape etc.
#                       # X3dToPython.xslt stylesheet insertPackagePrefix=true supports this option.
#
####################################################################################################

from x3d import *

newModel=X3D(profile='Immersive',version='3.3',
  head=head(
    children=[
    meta(content='LoadSensorExample.x3d',name='title'),
    meta(content='Simple test of LoadSensor node',name='description'),
    meta(content='Don Brutzman',name='creator'),
    meta(content='17 July 2008',name='created'),
    meta(content='14 June 2020',name='modified'),
    meta(content='Copyright Don Brutzman and Leonard Daly 2008',name='rights'),
    meta(content='X3D LoadSensor example',name='subject'),
    meta(content='https://X3dGraphics.com/examples/X3dForWebAuthors/Chapter12EnvironmentSensorSound/LoadSensorExample.x3d',name='identifier'),
    meta(content='X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit',name='generator'),
    meta(content='../license.html',name='license')]),
  Scene=Scene(
    children=[
    WorldInfo(title='LoadSensor.x3d'),
    Viewpoint(DEF='ViewAll',description='LoadSensor test',position=(0,0,12)),
    Inline(DEF='HelloWorld',url=["../HelloWorld.x3d","https://www.web3d.org/x3d/content/examples/HelloWorld.x3d"]),
    LoadSensor(DEF='InlineLoadSensor',timeOut=5,
      children=[
      Inline(USE='HelloWorld')]),
    Transform(DEF='PivotTextSigns',translation=(0,3,0),
      #  First sign indicates waiting... 
      children=[
      Shape(
        geometry=Text(length=[0],string=["LoadSensor waiting for HelloWorld scene..."],
          fontStyle=FontStyle(justify=["MIDDLE","MIDDLE"],size=0.7,style_='ITALIC')),
        appearance=Appearance(
          material=Material(emissiveColor=(0.878431,0.262745,0)))),
      #  Second sign indicates completed, initially rotated 90 degrees out of view 
      Transform(rotation=(0,1,0,1.57),
        children=[
        Shape(
          geometry=Text(length=[0],string=["LoadSensor reports HelloWorld scene loading complete."],
            fontStyle=FontStyle(justify=["MIDDLE","MIDDLE"],size=0.7,style_='ITALIC')),
          appearance=Appearance(
            material=Material(emissiveColor=(0,0.278431,0.027451))))])]),
    OrientationInterpolator(DEF='SignRotationInterpolator',key=[0,1],keyValue=[(0,1,0,0),(0,1,0,-1.57)]),
    TimeSensor(DEF='SignRotationClock',cycleInterval=0.8),
    ROUTE(fromField='loadTime',fromNode='InlineLoadSensor',toField='set_startTime',toNode='SignRotationClock'),
    ROUTE(fromField='fraction_changed',fromNode='SignRotationClock',toField='set_fraction',toNode='SignRotationInterpolator'),
    ROUTE(fromField='value_changed',fromNode='SignRotationInterpolator',toField='set_rotation',toNode='PivotTextSigns'),
    Script(DEF='TraceScript',
      field=[
      field(accessType='inputOnly',name='eventLoaded',type='SFBool'),
      field(accessType='inputOnly',name='loadTimeEvent',type='SFTime')],

    sourceCode="""
ecmascript:
function eventLoaded (value)
{
	Browser.println ('InlineLoadSensor isLoaded=' + value);
}
function loadTimeEvent (value)
{
	Browser.println ('InlineLoadSensor loadTime=' + value);
}
"""),
    ROUTE(fromField='isLoaded',fromNode='InlineLoadSensor',toField='eventLoaded',toNode='TraceScript'),
    ROUTE(fromField='loadTime',fromNode='InlineLoadSensor',toField='loadTimeEvent',toNode='TraceScript')])
) # X3D model complete

####################################################################################################
# Self-test diagnostics
####################################################################################################

print('Self-test diagnostics for LoadSensorExample.py:')
if        metaDiagnostics(newModel): # built-in utility method in X3D class
    print(metaDiagnostics(newModel)) # display meta info, hint, warning, error, TODO values in this model
# print('check newModel.XML() serialization...')
newModelXML= newModel.XML() # test export method XML() for exceptions during export
newModel.XMLvalidate()
# print(newModelXML) # diagnostic

try:
#   print('check newModel.VRML() serialization...')
    newModelVRML=newModel.VRML() # test export method VRML() for exceptions during export
    # print(prependLineNumbers(newModelVRML)) # debug
    print("Python-to-VRML export of VRML output successful", flush=True)
except Exception as err: # usually BaseException
    # https://stackoverflow.com/questions/18176602/how-to-get-the-name-of-an-exception-that-was-caught-in-python
    print("*** Python-to-VRML export of VRML output failed:", type(err).__name__, err)
    if newModelVRML: # may have failed to generate
        print(prependLineNumbers(newModelVRML, err.lineno))

try:
#   print('check newModel.JSON() serialization...')
    newModelJSON=newModel.JSON() # test export method JSON() for exceptions during export
#   print(prependLineNumbers(newModelJSON)) # debug
    print("Python-to-JSON export of JSON output successful (under development)")
except Exception as err: # usually SyntaxError
    print("*** Python-to-JSON export of JSON output failed:", type(err).__name__, err)
    if newModelJSON: # may have failed to generate
        print(prependLineNumbers(newModelJSON,err.lineno))

print("python LoadSensorExample.py load and self-test diagnostics complete.")
